### ISSUE TITLE
Help with understanding py_environment.PyEnvironment wrapping potential

### ISSUE THREAD

**Issue by jm-willy** [tensorflow/agents#837]
Hi! I created a python (no tf agents) chess environment for my RL training loop and I have some doubts. It's going to be full open source, but simplifying, the code looks like this:

Enviroment:
```
import chess # python-chess has the game logic, board representation and chess engine uci interface

class CustomChessEnv
    # this class converts a python-chess board string to input tensor
    # calculates rewards, checking if the actor-critic made an illegal move
    def do_stuff()
        return
```
Loop:
```
env = CustomChessEnv()
with tf.GradientTape() as tape:
    while loop:
        env.do_stuff()
```

The loop fails because env.do() does lots of python operations, like append, dict key search, flow control and a lot of arithmetic. Some, like append(and potentially anything dict related) seem to not be allowed within tf.GradientTape(), giving error saying that a keras tensor doesn't have an append method.

My questions are: 
- Can I inherit (py_environment.PyEnvironment) in a new class and use my own CustomChessEnv methods inside inherited py_environment.PyEnvironment's main interface methods (step, _step, etc) to wrap my logic? Should I do this through composition or (multiple) inheritance?

- Similar to above, are external modules like python-chess incompatible in some way? I use python-chess to interface with a chess engine

- What is the time_step_spec abstract method?

Thank you for any help

**Comment by jm-willy** [comment]
Gradient tape was recording a list which was converted to a tf.constant, worked after removal of tf.constant
